lib/galaxy/managers/hdcas.py

```diff
@@ -103,6 +108,23 @@ def update_attributes(self, content, payload: Dict):
         # pre-requisite checked that attributes are valid
         self.map_datasets(content, fn=lambda item, *args: set_collection_attributes(item, payload.items()))
 
+    def delete(self, item, flush=True, **kwargs):
+        super().delete(item, flush=flush, **kwargs)
+        self.map_datasets(item, fn=self._delete_hda_if_orphaned_and_hidden)
+
+    def _delete_hda_if_orphaned_and_hidden(self, dataset, *parents):
+        if isinstance(dataset, model.HistoryDatasetAssociation):
+            if not dataset.deleted and not dataset.purged and not dataset.visible:
+                if not self._is_hda_used_in_multiple_collections(dataset):
+                    super().delete(dataset, flush=False)
+
+    def _is_hda_used_in_multiple_collections(self, dataset: model.HistoryDatasetAssociation):
+        stmt = select(func.count(model.DatasetCollectionElement.id)).where(
+            model.DatasetCollectionElement.hda_id == dataset.id,
+        )
+        usage_count = self.session().execute(stmt).scalar()
+        return usage_count > 1
```

Comment:
Is this enough to check that we can mark the dataset as deleted?
I just realized that if we have 2 collections pointing to the same datasets and we delete one collection this will not mark the datasets as deleted (which is what we want) but, if we then delete the other collection, we are still not marking the datasets as deleted. Should we filter out those DCEs whose ancestor HDCA is already deleted for the usage count?

Reply:
This no longer applies